### ISSUE TITLE
Support loaders for script Parameters

### ISSUE THREAD

**Issue by gaborbernat** [argoproj-labs/hera#903]
**Is your feature request related to a problem? Please describe.**

For a fanin operation, the arguments are always passed in as `json.dumps`. We already support loaders for artifacts to map them back to a data structure, would be nice if this could work for parameters too:

```python
class Magic(BaseModel):
    files: list[str]


@script(constructor="runner")
def fan_out() -> Annotated[Magic, Parameter(name="response")]:
    return Magic(files=["1.ok"])

# today we have
@script(constructor="runner")
def fan_in(*, responses: list[str]) -> None:
    values: list[Magic] = [Magic.parse_raw(r) for r in responses]
    print(values)
```

**Describe the solution you'd like**
```python
# would be nice if we could have
@script(constructor="runner")
def fan_in(*, responses: Annoted[list[Magic], Parameter(loader=Magic)]) -> None:
    print(responses)
```

**Describe alternatives you've considered**
Keep using the transform objects in the first line of your script.

**Comment by elliotgunton** [comment]
Related: #858 and #827 - better parameter loading is definitely on the roadmap, thanks for the suggestion @gaborbernat!

**Comment by elliotgunton** [comment]
Consolidating into https://github.com/argoproj-labs/hera/issues/1166 - this will be the next feature to implement as the most requested over various issues. Thanks for the `Parameter(loader=Magic)` syntax suggestion, I think I'll be using it!

**Comment by elliotgunton** [comment]
> For a fanin operation, the arguments are always passed in as json.dumps

This was actually a bug fixed in Argo upstream, see #1146 for context/links
